language-plutus-core/src/Language/PlutusCore/Constant/Function.hs

```diff
@@ -51,3 +53,60 @@ insertDynamicBuiltinNameDefinition
 -- | Return the 'Type' of a 'TypedBuiltinName'.
 typeOfTypedBuiltinName :: TypedBuiltinName uni as r -> Type TyName uni ()
 typeOfTypedBuiltinName (TypedBuiltinName _ scheme) = typeSchemeToType scheme
+
+{- | A type to represent types of built-in functions in a convenient
+   form.  We can't use the Type because that's ambiguous: eg, a
+   function taking one int argument and returning a function of type
+   int -> int has the same Type as a function taking two integers and
+   returning an int (ie int -> int -> int).  TypeSchemes for builtins
+   contain the information we require, but are a bit too general,
+   allowing one to represent types that builtins aren't allowed to
+   have (builtins have to have types with all quantifiers at the
+   start, but TypeSchemes can represent types with quantifiers in the
+   middle).  The TypeComponents type represents exactly the types we
+   need for typechecking builtins, which simplifies typechecking.  The
+   typeComponentsOfTypeScheme function attempts to convert a
+   TypeScheme to a TypeComponents object.
+-}
+data TypeComponents uni = TypeComponents { tcTypeVars   :: [TyName]
+                                         , tcArgTypes   :: [Type TyName uni ()]
```

Comment:
I think I tried that as well, but it was a bit annoying either way (if you make a VarDecl, you have to take it apart sometimes to get the types out).  I think @effectfully suggested that he might re-work the TypeScheme stuff at some time in the future, so with luck all this stuff might go away.

Reply:
> Although then you have to come up with a name which is probably a pain.

We could use `arg_1`, `arg_2` etc, but generating names is scary, so it's nice to avoid that when possible.

>  I think @effectfully suggested that he might re-work the TypeScheme stuff at some time in the future, so with luck all this stuff might go away.

I'm not sure if there's a better way of doing type checking even with reworked `TypeScheme`.